ducktape/tests/runner.py

```diff
@@ -14,216 +14,153 @@
-            self.stop_testing = True
+        """Start a test runner client in a subprocess"""
+        proc = multiprocessing.Process(
+            target=run_client,
+            args=[
+                self.hostname,
+                self.port,
+                self.current_test_context.test_id,
+                self.current_test_context.logger_name,
+                self.current_test_context.results_dir,
+                self.session_context.debug,
+                self.session_context.max_parallel
+            ])
+        self.proc_list.append(proc)
+        proc.start()
+
+    def handle(self, event):
+        self.log(logging.DEBUG, str(event))
+
+        if event["event_type"] == ClientEventSupplier.READY:
+            self.handle_ready(event)
+        elif event["event_type"] in [ClientEventSupplier.RUNNING, ClientEventSupplier.SETTING_UP, ClientEventSupplier.TEARING_DOWN]:
+            self.handle_lifecycle(event)
+        elif event["event_type"] == ClientEventSupplier.FINISHED:
+            self.handle_finished(event)
+        elif event["event_type"] == ClientEventSupplier.LOG:
+            self.handle_log(event)
+        else:
+            raise RuntimeError("Received event with unknown event type: " + str(event))
 
-    def log(self, log_level, msg):
-        """Log to the service log and the test log of the current test."""
+    def handle_ready(self, event):
+        self.receiver.send(
+                self.event_response.ready(event, self.session_context, self.current_test_context, self.cluster))
 
-        if self.current_test is None:
-            msg = "%s: %s" % (self.who_am_i(), str(msg))
-            self.logger.log(log_level, msg)
-        else:
-            msg = "%s: %s: %s" % (self.who_am_i(), self.current_test_context.test_name, str(msg))
-            self.logger.log(log_level, msg)
-            self.current_test.logger.log(log_level, msg)
+        # Test is considered "active" as soon as we receive the ready request
+        self.active_tests[event["test_id"]] = event
 
+    def handle_log(self, event):
+        self.receiver.send(self.event_response.log(event))
+        self.log(event["log_level"], event["message"])
 
+    def handle_finished(self, event):
+        self.receiver.send(self.event_response.finished(event))
 
+        # Move this test from running to finished
+        del self.active_tests[event["test_id"]]
+        self.finished_tests[event["test_id"]] = event
+        self.results.append(event['result'])
 
+        if len(self.staged_tests) + len(self.active_tests) > 0 and self.session_context.max_parallel == 1:
+            terminal_width, y = get_terminal_size()
+            print "~" * int(2 * terminal_width / 3)
 
+    def handle_lifecycle(self, event):
+        self.receiver.send(self.event_response.event_response(event))
 
+    def log_event(self, log_level, event):
+        """Log an event received from the client."""
+        pass
+
+    def log(self, log_level, msg):
```

Comment:
Probably doesn't matter here as its just for this class, but if we're proxying for loggers, you might want to provide `*args` and `**kwargs` support as in the `logging` module (https://docs.python.org/3/library/logging.html#logging.Logger.log) since it doesn't require you to do the interpolation step yourself.